hpke/aead.go

```diff
@@ -3,21 +3,115 @@ package hpke
 import (
 	"crypto/cipher"
 	"errors"
+
+	"golang.org/x/crypto/cryptobyte"
 )
 
 type encdecCtx struct {
-	cipher.AEAD
-	Suite
+	raw []byte
+
+	// parsed from raw
+	suite          Suite
+	exporterSecret []byte
+	key            []byte
 	baseNonce      []byte
 	seq            []byte
-	exporterSecret []byte
+
+	// operational parameters
+	cipher.AEAD
 }
 
 type sealCtx struct{ *encdecCtx }
 type openCtx struct{ *encdecCtx }
 
 func (c *encdecCtx) Export(expCtx []byte, len uint16) []byte {
-	return c.labeledExpand(c.exporterSecret, []byte("sec"), expCtx, len)
+	return c.suite.labeledExpand(c.exporterSecret, []byte("sec"), expCtx, len)
+}
+
+func (c *encdecCtx) Suite() Suite {
+	return c.suite
+}
+
+// marshal serializes an HPKE context.
+func (c *encdecCtx) marshal() ([]byte, error) {
+	if c.raw != nil {
+		return c.raw, nil
+	}
+
+	var b cryptobyte.Builder
+	b.AddUint16(uint16(c.suite.KemID))
+	b.AddUint16(uint16(c.suite.KdfID))
+	b.AddUint16(uint16(c.suite.AeadID))
+	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes(c.exporterSecret)
+	})
+	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes(c.key)
+	})
+	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes(c.baseNonce)
+	})
+	b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes(c.seq)
+	})
+	return b.Bytes()
```

Comment:
the raw field must be updated with the result.

Reply:
Good catch! The fix is a little different though. I removed `raw` from `encdecCtx` and moved it to `sealCtx` and `openCtx`.